### PR TITLE
[trainer, cfg, tests] fix: abort on a sustained non-finite grad_norm streak

### PR DESCRIPTION
**Pull request by OnePunchMonk**
## Summary

Item A3 of #388's fail-fast checklist. Both the FSDP and VeOmni diffusion engines (`workers/engine/fsdp/diffusers_impl.py`, `workers/engine/veomni/diffusion_impl.py`) silently skip the optimizer step and only `print`/`logger.warning` on a non-finite `grad_norm`. If a run gets stuck producing non-finite gradients, every subsequent optimizer step is silently skipped forever — the run keeps burning GPU time with no visible failure signal, which is exactly the silent-fallback pattern #388 was filed to eliminate.

## What changed

- Added `track_nonfinite_grad_streak()` (`verl_omni/trainer/diffusion/diffusion_trainer_utils.py`), a small pure function that tracks the consecutive non-finite-`grad_norm` streak and raises once a configurable threshold is exceeded.
- Wired it into both training loops that already read `metrics.get("actor/grad_norm")`: `v1/trainer_base.py` (`_compute_metrics`) and `ray_diffusion_trainer.py` (`PolicyGradientRayTrainer.fit` and `DirectPreferenceRayTrainer.fit`). Each now exports `train/nonfinite_grad_steps` and aborts past the threshold.
- Added `trainer.max_consecutive_nonfinite_grad_steps` (default `0` = disabled, preserving current behavior) to `diffusion_trainer.yaml`; regenerated the derived `_generated_diffusion_trainer.yaml` / `_generated_diffusion_veomni_trainer.yaml`.
- Bumped both engines' silent `logger.warning`/`print` to `logger.error`.
- Added `tests/trainer/diffusion/test_nonfinite_grad_streak_on_cpu.py`.

## Why this is not duplicate work

Checked before starting:

```bash
gh issue view 388 --repo verl-project/verl-omni --comments
gh pr list --repo verl-project/verl-omni --search "grad_norm" --state all
gh pr list --repo verl-project/verl-omni --search "nonfinite" --state all
gh pr list --repo verl-project/verl-omni --search "diffusion_impl.py" --state all
```

The three PRs already merged against #388 (#443, #429, #497) do not touch `workers/engine/veomni/diffusion_impl.py` or `workers/engine/fsdp/diffusers_impl.py` — confirmed by diffing their changed-file lists. #429's own branch name (`wenzhe/fail-fast-a5-a8`) explicitly scopes it to items A5/A8, not A3. `main` still has the bare `logger.warning`/`print` with no counter as of this PR.

## Test plan

Ran on a Modal A10G GPU box (this dev machine cannot install the GPU-only `vllm-omni` wheel), following `docs/start/install.md`:

```bash
uv pip install -e ".[gpu]" --torch-backend=auto
uv pip install "vllm-omni @ git+https://github.com/vllm-project/vllm-omni.git@$(cat .github/vllm_omni_pin.txt)"
uv pip install -e ".[train,dev]"

python -m pytest -q tests/trainer/diffusion/test_nonfinite_grad_streak_on_cpu.py \
  tests/trainer/diffusion/test_diffusion_teacher_manager_on_cpu.py \
  tests/trainer/diffusion/test_diffusion_core_algos_on_cpu.py
# 44 passed

python -m pytest -q tests/trainer/diffusion
# 311 passed
```

Also ran `ruff check` / `ruff format --check` locally on every changed `.py` file (clean) and `git diff --check` (clean). Did not run `pre-commit` itself (not available on this dev machine) or `scripts/generate_trainer_config.sh` on the GPU box — the generated YAML diffs were produced locally via that script and match the field addition.

## AI assistance disclosure

- [x] This PR was developed with AI assistance (Claude). The submitting human reviewed the change end-to-end and can defend it.

- [x] A manual human reviewer (onepunchmonk) has reviewed the changes before marking them ready for review.

### CODE REVIEW

File: verl_omni/trainer/diffusion/ray_diffusion_trainer.py [hunk 1]
```diff
@@ -1880,6 +1889,12 @@ def fit(self):
                 if "advantages" in batch.batch:
                     gradient_norm = metrics.get("actor/grad_norm", None)
                     metrics.update(compute_variance_proxy_metrics(batch=batch, gradient_norm=gradient_norm))
+                    self._nonfinite_grad_streak = track_nonfinite_grad_streak(
+                        self._nonfinite_grad_streak,
+                        gradient_norm,
+                        self.config.trainer.get("max_consecutive_nonfinite_grad_steps", 0),
+                    )
+                    metrics["train/nonfinite_grad_steps"] = self._nonfinite_grad_streak
```
[the remark this answers is not in the document]
Comment (OnePunchMonk) [reply verl_omni/trainer/diffusion/ray_diffusion_trainer.py]: Good catch — fixed. Moved `gradient_norm` and the streak tracking out of the `if "advantages" in batch.batch` gate in both PolicyGradient's DirectPreference loop and the v1 trainer, so the abort/metric now runs regardless of whether `advantages` is present.

**Comment by OnePunchMonk** [comment]
@zhtmike / @SamitHuang, can one of you please provide a review here? Thank you!

**Comment by zhtmike** [comment]
Hi @OnePunchMonk , I think an occasional non-finite grad_norm might be reasonable, and a global grad clip is default applied. So A3 might be a false alarm. WDYT?

**Comment by OnePunchMonk** [comment]
Thanks for the pushback. Ran a quick check on this before replying.

Grad clipping doesn't actually cover this case. I tested `torch.nn.utils.clip_grad_norm_` directly with NaN and Inf gradients on CPU: it can't rescale a gradient that's already non-finite, it just returns the non-finite norm unchanged and skips scaling.

```
grad=nan: clip_grad_norm_ returns nan, no rescaling
grad=inf: clip_grad_norm_ returns inf, no rescaling
grad=5.0: clip_grad_norm_ returns 10.0, rescaled correctly
```

Feeding that straight into `optimizer.step()` turns every param to NaN. So the thing actually preventing corruption today is the existing `if not torch.isfinite(grad_norm): skip step` branch in `diffusers_impl.py`/`diffusion_impl.py`, not the clip.

That said, you're right that an occasional isolated non-finite step is normal and shouldn't be treated as a failure. This PR doesn't change that: the default threshold is `0` (disabled), so a single blip still gets skipped exactly like today, no behavior change. What it adds is an opt-in cap for the case where the skip repeats indefinitely, a run stuck producing non-finite grads forever, silently skipping every step with no forward progress and no signal, which is what #388 was about.

If you'd rather keep this off by default entirely and just ship the metric (`train/nonfinite_grad_steps`) for visibility without any abort path, I'm fine narrowing the PR to that. WDYT?

**Comment by zhtmike** [comment]
I need @knlnguyen1802 opinion for this. For my side, I think we follow verl's practice (a warning at training) is just enough.

File: verl_omni/trainer/diffusion/diffusion_trainer_utils.py [hunk 2]
```diff
@@ -66,6 +67,20 @@ def worker_group_port_ranges(master_port_range: Optional[Sequence[int]], num_gro
     return [[lo + i * stride, hi if i == num_groups - 1 else lo + (i + 1) * stride] for i in range(num_groups)]
 
 
+def track_nonfinite_grad_streak(streak: int, grad_norm: Optional[float], max_consecutive: int) -> int:
+    """Update the consecutive-non-finite-gradient streak, raising past ``max_consecutive`` (<= 0 disables)."""
+    if grad_norm is None or math.isfinite(grad_norm):
+        return 0
+    streak += 1
+    if max_consecutive > 0 and streak > max_consecutive:
```
Comment (wtomin) [inline verl_omni/trainer/diffusion/diffusion_trainer_utils.py:75]: Suggest to change it to `max_consecutive > 0 and streak >= max_consecutive`, so that it better aligns with the abortion message. Suppose `max_consecutive=3`, `streak = 3` should raise an error, instead of `streak = 4`.

**Comment by OnePunchMonk** [comment]
Fixed the off-by-one: `streak >= max_consecutive` now, so `max_consecutive=3` raises on the 3rd consecutive non-finite step. Updated the tests to match.

**Comment by knlnguyen1802** [comment]
I think we should follow verl practice 
In upstream they just warning and skip with continue
Please check the upstream behavior again

**Comment by OnePunchMonk** [comment]
Checked upstream verl directly (`verl/workers/engine/fsdp/transformer_impl.py` and `verl/workers/engine/veomni/transformer_impl.py`): confirmed, upstream today is exactly `if not torch.isfinite(grad_norm): print(f"WARN: grad_norm is not finite: {grad_norm}")` then skip the step, no counter, no abort. So you're right that upstream's baseline behavior is warn-and-skip only.

This PR doesn't change that baseline: `trainer.max_consecutive_nonfinite_grad_steps` defaults to `0`, which disables the abort entirely, so out of the box every run behaves identically to upstream (skip + log). The only default-on change is the two engine call sites going from `print`/`logger.warning` to `logger.error`, and the new `train/nonfinite_grad_steps` metric, which is visibility, not a behavior change.

The abort path is opt-in and exists for the case upstream doesn't handle at all: a run that gets stuck emitting non-finite gradients forever, which under warn-and-skip-only just burns GPU indefinitely with a wall of warnings and no way to fail the job automatically. That's the literal case #388 was filed for.

If the concern is scope rather than correctness, happy to narrow this PR to just the metric + log-level bump and drop the abort path, so the diff matches upstream's behavior exactly and the fail-fast piece can be revisited separately. Let me know which way you'd like to go and I'll adjust.

**Comment by wtomin** [comment]
@zhtmike @knlnguyen1802  Any comments for the author's last response? 

I think "narrowing this PR to just the metric + log-level " is a good idea. `max_consecutive_nonfinite_grad_steps` is not necessary.

**Comment by zhtmike** [comment]
> @zhtmike @knlnguyen1802 Any comments for the author's last response?
> 
> I think "narrowing this PR to just the metric + log-level " is a good idea. `max_consecutive_nonfinite_grad_steps` is not necessary.

yes, pls do so.

**Comment by OnePunchMonk** [comment]
Narrowed as discussed: dropped \`trainer.max_consecutive_nonfinite_grad_steps\` and the abort/raise entirely. What's left:

- \`track_nonfinite_grad_streak\` now just tracks the streak for visibility, no threshold, no raise.
- \`train/nonfinite_grad_steps\` metric, wired in the same three call sites (unaffected by the earlier "advantages"-gating fix, which still stands).
- The \`logger.error\` bump on the two engine call sites (was \`print\`/\`logger.warning\`).

Diff against main is now 6 files / +64 -4. Test file trimmed to match (dropped the threshold/raise cases). PTAL.

**Comment by wtomin** [comment]
I am ok with tracking non_finite grads steps, but I am not sure if changing a `warning` message to an error message is good. Also cc @zhtmike .

**Comment by OnePunchMonk** [comment]
Reverted the warning->error bump — kept it at `logger.warning` since it's the existing skip-the-step path, not a new failure mode. Pushed.
